apps/designer/app/canvas/shared/styles.ts

```diff
@@ -4,59 +4,95 @@ import { setInstanceStyleMutable } from "~/shared/tree-utils";
+import { FontFormat, FONT_FORMATS, getFontFaces } from "@webstudio-is/fonts";
+
+const cssEngine = createCssEngine();
 
 const voidElements =
   "area, base, br, col, embed, hr, img, input, link, meta, source, track, wbr";
 
 // Helper styles on for canvas in design mode
-const styles = {
-  "[data-ws-component]": {
-    // When double clicking into an element to edit text, it should not select the word.
-    userSelect: "none",
-  },
-  [`[data-ws-component]:not(${voidElements}):not(body):empty`]: {
-    outline: "1px dashed #555",
-    outlineOffset: -1,
-    paddingTop: 50,
-    paddingRight: 50,
-  },
-  "[data-ws-component][contenteditable], [data-ws-component]:focus": {
-    outline: 0,
-  },
-  "[data-ws-component][contenteditable]": {
-    boxShadow: "0 0 0px 4px rgb(36 150 255 / 20%)",
-  },
+const helperStyles = [
+  // When double clicking into an element to edit text, it should not select the word.
+  `[${idAttribute}] {
+    user-select: none;
+  }`,
+  `[${idAttribute}]:not(${voidElements}):not(body):empty {
+    outline: 1px dashed #555;
+    outline-offset: -1;
+    padding-top: 50;
+    padding-right: 50;
+  }`,
+  `[${idAttribute}][contenteditable], [${idAttribute}]:focus {
+    outline: 0;
+  }`,
+  `[${idAttribute}][contenteditable] {
+    box-shadow: 0 0 0px 4px rgb(36 150 255 / 20%)
+  }`,
   // Text Editor wraps each line into a p, so we need to make sure there is no jump between regular rendering and editing
-  "[data-ws-component][contenteditable] p": {
-    margin: 0,
-  },
-};
-
-// @todo rewrite to use css engine and remove stitches from canvas entirely
-const wrapperComponentGlobalStyles = globalCss(styles);
+  `[${idAttribute}][contenteditable] p {
+    margin: 0
+  }`,
+];
 
-export const useManageStyles = () => {
-  wrapperComponentGlobalStyles();
+export const useManageDesignModeStyles = () => {
   useUpdateStyle();
   usePreviewStyle();
   useRemoveSsrStyles();
 };
 
-const cssEngine = createCssEngine();
+export const addGlobalRules = (
+  engine: CssEngine,
+  { assets = [] }: { assets?: Array<Asset> }
+) => {
+  // @todo we need to figure out all global resets while keeping
+  // the engine aware of all of them.
+  // Ideally, the user is somehow aware and in control of the reset
+  engine.addPlaintextRule("html {margin: 0; height: 100%}");
+
+  const fontAssets = assets.filter((asset) =>
+    FONT_FORMATS.has(asset.format as FontFormat)
+  ) as Array<FontAsset>;
+  const fontFaces = getFontFaces(fontAssets);
+  for (const fontFace of fontFaces) {
+    engine.addFontFaceRule(fontFace);
+  }
+};
+
+const globalStylesCssEngine = createCssEngine();
+
+export const useGlobalStyles = (assets: Array<Asset>) => {
+  globalStylesCssEngine.clear();
+
+  addGlobalRules(globalStylesCssEngine, { assets });
+
+  for (const style of helperStyles) {
+    globalStylesCssEngine.addPlaintextRule(style);
+  }
+
+  if (typeof document !== "undefined") {
+    globalStylesCssEngine.render();
+  }
```

Comment:
Seems odd that we do all these side effects during render rather than in useEffect. If there's a reason for this, I think we need a code comment.

Reply:
true I took a bad shortcut, because I need it to be rendered first, before all other components styles get rendered

Reply:
Another option is to render a component first and then call this hook inside it and use layout effect right away. So that it renders as early as possible

Reply:
Sounds like use-case for useLayoutEffect, yeah